File: player-mfe/src/entities/chunk.ts
import { ChunkResponse } from "../types/requests";
import Frame from "./frame";

class Chunk {
  private frames: Frame[];
  private lastFrame: Frame | null;
  private requestTime: Date;
  private endTime: Date;

  constructor(data: ChunkResponse) {
    this.frames = data.frames.map(
      (frame) => new Frame(frame, Date.parse(data.end_time))
    );
    this.lastFrame = null;
    this.requestTime = new Date(data.request_time);
    this.endTime = new Date(data.end_time);
  }

  public nextFrame() {
    this.lastFrame = this.frames.shift() ?? null;
    return this.lastFrame;
  }

  public getLastFrame() {
    return this.lastFrame;
  }

  public getRequestTime() {
    return this.requestTime;
  }

  public getEndTime() {
    return this.endTime;
  }

  public getLastTimestamp() {
    return this.frames[this.frames.length - 1]?.getCurrentTimestamp();
  }
}

export default Chunk;

File: player-bff/player-sonson/internal/entity/chunk.go
package entity

import (
	"encoding/json"
	"time"
)

type Chunk struct {
	RequestTime time.Time `json:"request_time"`
	EndTime     time.Time `json:"end_time"`
	Frames      []Frame   `json:"frames"`
}

func (c *Chunk) ToJson() ([]byte, error) {
	return json.Marshal(c)
}
